TwentyTwentyOne: Restructure button style cascade to improve button color options.

allancole opened the pull request:
<!-- Add reference to the issue this pull-request fixes.-->
Fixes #858
_Possibly_ fixes: #880

## Summary
Continues with @carolinan work in #879 so im borrowing the summary from there. The cover block sets the links to currentColor so that the text and link color has a high contrast against a chosen palette color. This style was overriding the button block and file block link colors. 

Additionally, this PR attempts to address all Blocks that use a button and use the same border-width style consistently. 
## Test instructions

This PR can be tested by following these steps:

1. Create a post with this [sample content](https://cloudup.com/cXTzNhag6EV)
2. Save and view the front
3. Compare the colors and confirm that all the text is readable.
4. Test hover and focus styles.
5. Turn Dark Mode on or off, test both body background colors.
6. Repeat step 5 with the Block editor.

## Screenshots (UPDATED: 2020-11-25)

Before | After | Before (darkmode) | After (darkmode) 
------------|------------|------------|------------
![image](https://user-images.githubusercontent.com/709581/100164283-8f9ea900-2e85-11eb-99b4-102cd0aad509.png)|![image](https://user-images.githubusercontent.com/709581/100289802-c0471700-2f47-11eb-8fed-e5bc559955bc.png)|![image](https://user-images.githubusercontent.com/709581/100164208-69790900-2e85-11eb-8605-909b605050b2.png)|![image](https://user-images.githubusercontent.com/709581/100289535-22534c80-2f47-11eb-85ac-a3bfc1f5d7d0.png)

## Todo

- [x] Finish solution for nested button colors on a lighter background color.
- [ ] Test and apply this fix to the editor.

## Quality assurance
* [x] I have thought about any security implications this code might add.
* [x] I have checked that this code doesn't impact performance (greatly).
* [x] I have tested this code to the best of my abilities
* [ ] I have checked that this code does not affect the accessibility negatively

allancole committed:
TwentyTwentyOne: Set border color to transparent by default.

allancole committed:
TwentyTwentyOne: Revise cascading for `.has-background` and `.is-dark-theme`

allancole committed:
TwentyTwentyOne: Recompile.

allancole committed:
TwentyTwentyOne: Prevent file button from using currentColor in certian conditions.

allancole committed:
TwentyTwentyOne: Improve has-background colors for file block.

allancole committed:
TwentyTwentyOne: Recompile.

carolinan commented:
I did not understand why ` !important` is needed for the search block? Or if there was some mismatch with the previous padding?

carolinan commented:
Also we need to watch the search form (not block) and search widget submit button, it now has no visible border.

allancole commented:
> I did not understand why !important is needed for the search block? Or if there was some mismatch with the previous padding?

Yeah, I still need to audit much of the changes I’ve made here. There’s quite a few `!important` rules that can probably be removed. 

Additionally, this is likely to turn into quite a large refactor as I work on it more. Ideally we’d want the majority of the button style rules to live in the `button-style()` mixin so we’re sure all buttons follow the same standard behaviors. This makes the block-specific overrides easier to deal with and we’ll hopefullly end up with less CSS lines overall in the end. 

> Also we need to watch the search form (not block) and search widget submit button, it now has no visible border.

Ahh, good catch. Should be easy to attach the widget submit button to the button mixin. Widget form borders should be pretty straightforward too.

ryelle commented:
Thanks for the really comprehensive test content 💯 

Here are a few issues I caught running through this - I see you're still working on the light-group issues so I left those out.

The "outline" button with a background color has no border (both custom and palette colors):
![light-mode-no-bg-link](https://user-images.githubusercontent.com/541093/100255659-c0c4bb00-2f11-11eb-9bea-e3fbde948f34.png)

It's more obvious on focus:
![light-mode-no-bg-focus](https://user-images.githubusercontent.com/541093/100255661-c0c4bb00-2f11-11eb-816e-0190a4b889c5.png)

In the focus states for outline buttons on dark group backgrounds (light site background), only the default button has a white focus, the rest are dark
![light-mode-dark-section-focus](https://user-images.githubusercontent.com/541093/100255662-c0c4bb00-2f11-11eb-9e72-a69712201f3d.png)

> Ideally we’d want the majority of the button style rules to live in the `button-style()` mixin so we’re sure all buttons follow the same standard behaviors.

This sounds great, it would cut down on a lot of these edge cases if it can work out 🙂

carolinan commented:
The buttons _are_ different though, they do not follow the same behavior, only the styles that are used by all of them should be in the mixin?

allancole committed:
TwentyTwentyOne: Refactor how button styles cascade through the system.
- Moves most all button style rules to the mixin.
- Restrict custom rules and overrides to individual blocks with button or special button elements.

allancole committed:
TwentyTwentyOne: Recompile.

allancole committed:
TwentyTwentyOne: Add dark theme rules to `inside` search button option.

allancole committed:
TwentyTwentyOne: Recompile.

allancole committed:
TwentyTwentyOne: Adjust widget search button styles, code tidying.

allancole committed:
TwentyTwentyOne: recompile.

allancole commented:
> The buttons are different though, they do not follow the same behavior, only the styles that are used by all of them should be in the mixin?

@carolinan, If i’m understanding you correctly here, yes but what I’m proposing is we establish the default button style in the `button-style()` mixin as the _single source of truth_. Then, when we need to add tweaks for buttons in different contexts (button block, search form blocks, widgets, comment form etc.), we can do context-specific overrides in each relevant Sass partial. Let me know if that makes sense. The last few commits make for a better example of what I mean where we get something closer to what users expect, but also with less lines of overriding CSS :-)

allancole commented:
@ryelle, I think those last few commits should solve all three of the issues you found plus fixes the ”light-group“ background issue :-)

allancole committed:
TwentyTwentyOne: Code tidying.

allancole committed:
TwentyTwentyOne: Code tidying.

allancole committed:
TwentyTwentyOne: Clean up darkmode overrides when combined with nested background-colors.

allancole committed:
TwentyTwentyOne: Recompile.

allancole committed:
TwentyTwentyOne: Code tidying.

carolinan commented:
404 page search form: Button is missing border when focused.
Widget search form: Button is missing border when focused.

-----
Search block (default body background color)

Default, Button is missing border except on hover. On hover: Black text over grey background is unreadable.
Button inside:  Button is missing border except on hover. On hover: Black text over grey background is unreadable.
Button inside, icon:  Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
Button only, with icon:  Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
Button only, without icon:  Button is missing border except on hover. On hover: Black icon over grey background is unreadable.

Search block (dark mode enabled)
Button inside: On hover: The minimal color change is not different enough from the default state (accessibility). 
                       On focus: Text and icon are very light grey over white background, the contrast is not high enough (accessibility).

carolinan commented:
Form submit button, for example the comment form: Border is missing on focus.

carolinan commented:
On default body background, in the editor:
On button creation, only the dotted outline is visible. 

On hover: Outlined buttons without color settings have grey background and dark grey text.

-----

On dark body backgrounds (not dark mode), in the editor:
On button creation, only the dotted outline is visible. 

The placeholder text is not visible until the button is hovered over.
On hover: Outlined buttons without color settings have white background and white text.

----
With dark mode enabled, in the editor:
On button creation, only the dotted outline is visible. 
The placeholder text is white on white background
On hover: Outlined buttons without color settings have white background and white text.

-----


Because the text is white on white background when hovered, you can't see what you are typing inside the button unless you first move the mouse.

carolinan commented:
On the default body background:

Buttons placed in a block that has the grey palette color as background are invisible, only the text shows.
**This issue exists with and without this PR.**
On any other light background, the button background turns black and is visible.


default (filled) button on the default background, on hover (yup)
![default](https://user-images.githubusercontent.com/7422055/100602575-57272100-3304-11eb-8829-10ca1f8f2061.png)

With pink (sorry, red) body background, default (filled) button, on hover:
![pink](https://user-images.githubusercontent.com/7422055/100602444-2e9f2700-3304-11eb-88c2-9e729f2e6577.png)

Here, the contrast is too low to be accessible, but low contrast is better than invisible.

allancole committed:
Merge branch 'trunk' into try/restructure-button-style-cascade

allancole committed:
TwentyTwentyOne: Fix outline on focused buttons in widgets and 404 page.

allancole committed:
TwentyTwentyOne: Recompile.

allancole committed:
Merge branch 'try/restructure-button-style-cascade' of github.com:WordPress/twentytwentyone into try/restructure-button-style-cascade

allancole commented:
> 404 page search form: Button is missing border when focused.
> Widget search form: Button is missing border when focused.

Should be fixed in the latest commit :-) 

> **Search block (default body background color)**
> Default, Button is missing border except on hover. On hover: Black text over grey background is unreadable.
> Button inside: Button is missing border except on hover. On hover: Black text over grey background is unreadable.
> Button inside, icon: Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
> Button only, with icon: Button is missing border except on hover. On hover: Black icon over grey background is unreadable.
> Button only, without icon: Button is missing border except on hover. On hover: Black icon over grey background is unreadable.

Here what I see on this PR for each of those options you listed: 

![2020-11-30 17 16 16](https://user-images.githubusercontent.com/709581/100672334-ca925800-332f-11eb-83e6-cbe337c3159d.gif)

Could you share a screenshot of what you see @carolinan ? 

> **Search block (dark mode enabled)**
> Button inside: On hover: The minimal color change is not different enough from the default state (accessibility).
> On focus: Text and icon are very light grey over white background, the contrast is not high enough (accessibility).

This I do see ;-). Will try to work out a solution shortly.

carolinan commented:
Search block:
The border still disappears when I click on the buttons.
The color of the "button inside" is still incorrect when I click the button.

PC, win 10, Chrome version 87.0.4280.66. Gutenberg is not active.


![search-click](https://user-images.githubusercontent.com/7422055/100745009-f1f42e00-33de-11eb-911a-cd6513b75103.gif)

Compared with trunk:

![search-trunk](https://user-images.githubusercontent.com/7422055/100745260-4f887a80-33df-11eb-939e-9a9455657a29.gif)

carolinan commented on assets/sass/05-blocks/search/_style.scss:
```diff
@@ -52,12 +52,26 @@
+			border-color: transparent;
+		}
+
+		&:focus {
```
I think that removing this  focus style solves https://github.com/WordPress/twentytwentyone/pull/886#issuecomment-736542156 ?

kjellr replied:
Should be working better with 27f1fe1 + dbf60c7:

![Kapture 2020-12-01 at 09 52 19](https://user-images.githubusercontent.com/1202812/100756184-f9a1db80-33ba-11eb-90d4-b137364a3ffc.gif)

kjellr committed:
Use correct focus outer border color for search block buttons.

carolinan commented:
Front: 
With dark mode enabled, the black link background for focused links affects the filled button. But not the outline button.
![darkmode](https://user-images.githubusercontent.com/7422055/100755496-8913b280-33ec-11eb-9614-c9b090bef809.gif)

kjellr committed:
Fix inset search button text focus color.

kjellr commented:
The front-end seems to be mostly working in my testing (just using the sample data above). But I'm seeing some really glaring issues in the editor still: 

The outline button doesn't have readable hover text by default:

![Kapture 2020-12-01 at 09 55 21](https://user-images.githubusercontent.com/1202812/100756635-6ae18e80-33bb-11eb-88f2-5a19bd3b015b.gif)

And all of the search buttons appear incorrect: 

![Screen Shot 2020-12-01 at 9 54 59 AM](https://user-images.githubusercontent.com/1202812/100756682-79c84100-33bb-11eb-98f8-dc6280b59309.png)

carolinan commented:
Front:

Dark background, 
Search block with button inside, the button is invisible on focus:

![image](https://user-images.githubusercontent.com/7422055/100756782-de9c8f00-33ed-11eb-9707-d35356651918.png)

Dark mode:
Search block with button inside, the button is almost invisible on focus:
![image](https://user-images.githubusercontent.com/7422055/100757125-48b53400-33ee-11eb-9b28-d5da3e53b20f.png)

carolinan commented:
Front:
When I **click** on an outlined button, the border is invisible:
![image](https://user-images.githubusercontent.com/7422055/100757520-c2e5b880-33ee-11eb-9e1f-77a86318a79b.png)

kjellr committed:
Merge branch 'trunk' into try/restructure-button-style-cascade

kjellr committed:
Recompile CSS

allancole committed:
TwentyTwentyOne: recompile.

allancole committed:
Merge branch 'try/restructure-button-style-cascade' of github.com:WordPress/twentytwentyone into try/restructure-button-style-cascade

kjellr committed:
Fix front-end focus/active state issues for the outline block. Fix dark mode focus styles.

allancole committed:
Merge branch 'try/restructure-button-style-cascade' of github.com:WordPress/twentytwentyone into try/restructure-button-style-cascade
